ch8/LIS_BinarySearch.py

```diff
@@ -0,0 +1,35 @@
+# binary search method from JMB p.234-235
+
+def lst4(num, num_lis):
+    if num_lis == 0 or (num_lis > 0 and cache[num_lis - 1] <= num):
+        cache[num_lis] = num
+        num_lis += 1
+        return num_lis
+
+    lo = 0
+    hi = num_lis - 1
+    
+    # binary search
```

Comment:
"Although the basic idea of binary search is comparatively straightforward, the details can be surprisingly tricky" - Donald Knuth

"there is Bentley's experience that when he assigned binary search in courses for professional programmers at places like Bell Labs and IBM and gave them two hours, everyone reported they had got it right, and on examining their code 90% of them had bugs — year after year"

Binary search is one of those infamous algorithms that are so easy to come up with a 99% correct solution for and yet so hard to implement 100% correctly. Particularly, (not exhaustive) sources of errors include 1. off by one errors and 2. handling of duplicate items. 

Although it's hard to implement, the algorithm is very useful because 1. it is very fast (not so many other known logarithmic runtime algorithms) 2. the only requirement is that the input data may be sorted according to some standard (e.g. alphanumerically) so it is easily applicable. That said, I suggest this extreme idea: memorize some binary template code and stick with it. JMB includes a very good template of binary search and explains how to reason about its correctness in terms of invariants (i.e. statements that always hold true) (on page 131). This is personally my choice of the binary search template and I recommend it. 

On a similar note, thinking about your code in terms of its semantics is very important. You can think of your code as much more than lines of correct sequences of symbols according to some language. Rather, you can assign meaning to variables, lines or blocks of your code. For example, you can take lo, hi values to mean that "the answer to your binary search method is always in interval [lo, hi]." So it would probably be wrong to assign lo = 1 (because the answer may be 0, try submitting on algospot). but hi = num_lis (instead of num_lis - 1) would probably be ok according to your semantic (because the interval would always include the answer and your code would correctly tighten the interval to eventually find the answer). Note it happens to be that this meaning, or rather a statement, is always true. This can be your invariant. 

Forming correct semantics of your (complicated yet critical) code and coming up with corresponding invariants will serve you a long way. This is at least one trick many experienced competition programmers and/or great software engineers use to seemingly always implement tricky solutions correctly at one try. Assigning meaning or roles to blocks or files of code is also a trick to understand large scale (e.g. enterprise) modules and libraries because it is so easy to get lost when you try to understand a huge codebase line by line. Btw, an interval subfield of computer science and related topics exists and it is "formal methods"